catalog/CatalogDatabaseCache.hpp

```diff
@@ -0,0 +1,124 @@
+/**
+ *   Copyright 2015-2016 Pivotal Software, Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_CATALOG_CATALOG_DATABASE_CACHE_HPP_
+#define QUICKSTEP_CATALOG_CATALOG_DATABASE_CACHE_HPP_
+
+#include <cstddef>
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+#include "catalog/CatalogDatabaseLite.hpp"
+#include "catalog/CatalogTypedefs.hpp"
+#include "storage/StorageConstants.hpp"
+#include "threading/SharedMutex.hpp"
+#include "threading/SpinSharedMutex.hpp"
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+class CatalogRelationSchema;
+
+namespace serialization { class CatalogDatabase; }
+
+/** \addtogroup Catalog
+ *  @{
+ */
+
+/**
+ * @brief A database cache used in the distributed version.
```

Comment:
Quick description of the role of the cache and expected usage here would be great (can be refined over time) so that it is easier for a reader to understand how it fits in the overall architecture. Who uses it? When does a cache entry get invalidate? ...